doc/drop_counters/drop_counters_HLD.md

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+}
+```
+
+### 3.2.2 DEBUG_COUNTER_DROP_REASON Table
+Example:
+```
+{
+    "DEBUG_COUNTER_DROP_REASON": {
+        "DEBUG_0|SMAC_EQUALS_DMAC": {},
+        "DEBUG_0|INGRESS_VLAN_FILTER": {},
+        "DEBUG_1|EGRESS_VLAN_FILTER": {},
+        "DEBUG_2|TTL": {},
+    }
+}
+```
+
+## 3.3 State DB
+State DB will store information about:
+* What types of drop counters are available on this device
+* How many drop counters are available on this device
+* What drop reasons are supported by this device
+
+This information will be populated by the orchestrator (described later) on startup.
+
+### 3.3.1 SAI APIs
+We will use the following SAI APIs to get this information:
+* `sai_query_attribute_enum_values_capability` to query support for different types of counters
+* `sai_object_type_get_availability` to query the amount of available debug counters
+
+## 3.4 Counters DB
+The contents of the drop counters will be added to Counters DB by flex counters.
+
+## 3.5 SWSS
+A new orchestrator will be created to handle debug counter creation and configuration. Specifically, this orchestrator will support:
+* Creating a new counter
+* Deleting existing counters
+* Adding drop reasons to an existing counter
+* Removing a drop reason from a counter
+
+### 3.5.1 SAI APIs
+This orchestrator will interact with the following SAI Debug Counter APIs:
+* `sai_create_debug_counter_fn` to create/configure new drop counters.
+* `sai_remove_debug_counter_fn` to delete/free up drop counters that are no longer being used.
+* `sai_get_debug_counter_attribute_fn` to gather information about counters that have been configured (e.g. index, drop reasons, etc.).
+* `sai_set_debug_counter_attribute_fn` to re-configure drop reasons for counters that have already been created.
+
+## 3.6 syncd
+Flex counter will be extended to support switch-level SAI counters.
+
+# 4 Flows
+## 4.1 General Flow
+![alt text](./drop_counters_general_flow.png)
+The overall workflow is shown above in figure 1.
+
+(1) Users configure drop counters using the CLI. Configurations are stored in the DEBUG_COUNTER Config DB table.
+
+(2) The debug counts orchagent subscribes to the Config DB table. Once the configuration changes, the orchagent uses the debug SAI API to configure the drop counters.
+
+(3) The debug counts orchagent publishes counter configurations to Flex Counter DB.
+
+(4) Syncd subscribes to Flex Counter DB and sets up flex counters. Flex counters periodically query ASIC counters and publishes data to Counters DB.
+
+(5) CLI uses counters DB to satisfy CLI requests.
+
+(6) (not shown) CLI uses State DB to display hardware capabilities (e.g. how many counters are available, supported drop reasons, etc.)
+
+# 5 Warm Reboot Support
+At this stage, debug counters will be deleted prior to warm reboot and re-installed when orchagent starts back up. This is intended to simplify upgrade behavior and conserve hardware resources during the warm reboot.
```

Comment:
Do you preserve the drop counters values on warm reboots?

Reply:
I've added the caveat mentioned in the meeting yesterday. We will only remove the debug counters during warm reboot on platforms where the additional hardware resources are necessary.